### ISSUE TITLE
Process tracking can get off-by-one

### ISSUE THREAD

**Issue by arya** [cosmtrek/air#94]
I've observed that sometimes `air` fails to kill the previous version of the process and tries to start a new one and fails. Here is a log with the debug flag on:

```
make[1]: Entering directory '/app/go'
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": WRITE
api-server/app/auth/auth.go has changed
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": WRITE
api-server/app/auth/auth.go has changed
flushing events
api-server/app/auth/auth.go has changed
make[1]: Leaving directory '/app/go'
running...
[GIN-debug] Listening and serving HTTP on :8080
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": WRITE
api-server/app/auth/auth.go has changed
api-server/app/auth/auth.go has changed
building...
trying to kill cmd [/bin/sh -c bin/api-server]
cmd killed, pid: 274
make[1]: Entering directory '/app/go'
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": WRITE
api-server/app/auth/auth.go has changed
event: "/app/go/api-server/app/auth/auth.go": CHMOD
event: "/app/go/api-server/app/auth/auth.go": WRITE
api-server/app/auth/auth.go has changed
flushing events
api-server/app/auth/auth.go has changed
building...
make[1]: Entering directory '/app/go'
make[1]: Leaving directory '/app/go'
running...
[GIN-debug] Listening and serving HTTP on :8080
make[1]: Leaving directory '/app/go'
running...
[GIN-debug] Listening and serving HTTP on :8080
[GIN-debug] [ERROR] listen tcp :8080: bind: address already in use
```

**Comment by jhartman86** [comment]
Major +1 for this issue. I see this issue _all the time_, to the point where AIR is basically unusable and requires stopping/restarting the air process entirely... at which point, what's the point.

**Comment by xiantang** [comment]
> Major +1 for this issue. I see this issue _all the time_, to the point where AIR is basically unusable and requires stopping/restarting the air process entirely... at which point, what's the point.

can you provider more informations about this issue. such as os platform

**Comment by jhartman86** [comment]
@xiantang can't speak for original poster, but I see this on Alpine Linux (ie. inside a Docker container) all the time. It'll work for about 10 or so minutes before AIR seems like it loses track of which processes are running or not, then starts trying to build/run the latest prior to the other process having fully shutdown (or trying to shutdown at all).

**Comment by arya** [comment]
I'm on macOS running docker.

**Comment by naseem-opal** [comment]
I am also seeing this issue, but more or less immediately on MacOS not using Docker.

**Comment by technicallyjosh** [comment]
I ran into this issue all the time on OSX inside docker-compose. I fixed it by adding the following to a `.air.toml` file...

```toml
[build]
kill_delay = 1000
```

**Comment by silverwind** [comment]
Happens regularily here as well. Sometimes, when saving files in quick succession, Air seems to fail to kill of the old process which then get orphaned and the only remedy is to stop watching and `kill` it.

**Comment by Dids** [comment]
Could this potentially be remedied by always having the `delay` option be greated than `kill_delay`, to ensure that the process is killed before even attempting to trigger a new build?

In my case I'm relying on `send_interrupt` instead, so `kill_delay` should ideally never kick in, but I'm still seeing all sorts of weird behavior, such as the process getting killed before it has gracefully shut down, even though the `kill_delay` hasn't been reached yet.
My theory regarding this particular issue is that it's caused by new changes/builds triggering in quick succession, while the "kill timer" is still running for the original process, which ends up killing the wrong process effectively.

**Comment by mrsufgi** [comment]
happening for me too, but off by two (2), Mac M1pro within a docker container with dlv

**Comment by xiantang** [comment]
fixed in v1.29.0
